### PR TITLE
make queue re runnable

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
Please provide a description for what this PR is for.

And to help us figure out who should review this PR, please 
put an X in all the areas that this PR affects.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

**Comment by hzxuzhonghu** [comment]
The goal is to make some controller, like `namespace controller` rerunnable, so no need to reinitiate a new controller in leader election callback

File: pkg/queue/instance.go [hunk 1]
```diff
@@ -58,7 +59,15 @@ func (q *queueImpl) Push(item Task) {
 	q.cond.Signal()
 }
 
+// Run start the task handler, which should be in a separate go routine.
+// The queue can be closed via stop channel. However when it is being closed,
+// the left elements in the queue will be processed.
+// Note: The queue can be rerun after closed.
 func (q *queueImpl) Run(stop <-chan struct{}) {
+	// enable rerun the queue
```
Comment (howardjohn) [inline pkg/queue/instance.go:67]: What if they call run between `<-stop` and `q.cond.L.Lock()`? I think it may still end up being closed?

Comment (hzxuzhonghu) [reply pkg/queue/instance.go:67]: make sense, it may cause race now

Comment (hzxuzhonghu) [reply pkg/queue/instance.go:88]: done

File: pkg/queue/instance.go [hunk 2]
```diff
@@ -58,7 +65,30 @@ func (q *queueImpl) Push(item Task) {
...
+		q.mutex.Lock()
+		// set closed status
+		q.closed = true
+		q.mutex.Unlock()
+	}()
+
+	// enable rerun the queue
+	q.cond.L.Lock()
+	q.closing = false
+	q.cond.L.Unlock()
```
Comment (shalk) [inline pkg/queue/instance.go:91]: I think we can reset `q.closing`  when  shutdown queue at  `if len(q.tasks) == 0 `

**Comment by hzxuzhonghu** [comment]
This is not needed

**Pull request closed without merging**